<h1>Data Cleaning / Munging</h1>

<p>In these notes we look at creating various plots based on a set of data.  We have a datafile called degreeClassification.csv (available on blackboard) which consists of the following columns:</p>
<ul>
<li>Academic Year:  The year that the degree was awarded.</li>
<li>Mode of study: Full-time, part-time or both.</li>
<li>Sex: Male, Female, Other or all.</li>
<li>Country of HE provider: England, Northern Ireland, Scotland, Wales or all.</li>
<li>Classification of First Degree: First, Upper Second, Lower Second or Third/Pass.</li>
<li>Number: The number of students achieving that degree.</li>
<li>Percentage:  The percentage of students achieving that degree.</li>
</ul>   

<p>We start the exercises by getting the data ready to be plotted</p>

<h2>Import PANDAs</h2>

In [ ]:
pip install pandas


In [48]:
import pandas as pd

<h2>Read the data into a PANDAs DataFrame</h2>
<p>Read the 'degreeClassification.csv' in to a data frame called df, use the head method to output the first 20 rows of the datafile so that the structure of the dataframe can be explored.</p>

In [49]:
df = pd.read_csv('degreeClassification.csv')

In [50]:
df.head(20)

Title  \
0                   Location   
1             Academic years   
2                Data source   
3           Data source link   
4   Data file canonical link   
5                    Licence   
6                  Code page   
7                 Disclaimer   
8              2013/14 total   
9              2014/15 total   
10             2015/16 total   
11             2016/17 total   
12             2017/18 total   
13              Last updated   
14                       NaN   
15                       NaN   
16             Academic Year   
17                   2013/14   
18                   2013/14   
19                   2013/14   

   Percentage of first degree qualifiers obtaining each classification by sex, mode of study and country of HE provider  \
0                                                  UK                                                                     
1                                  2013/14 to 2017/18                                                                     
2                                                HESA                                                                     
3   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
4   https://www.hesa.ac.uk/data-and-analysis/sb252...                                                                     
5   Creative Commons Attribution 4.0 International...                                                                     
6                                      Unicode UTF-8                                                                      
7   Please note that this data includes rounded to...                                                                     
8                                              395605                                                                     
9                                              370905                                                                     
10                                             376330                                                                     
11                                             391875                                                                     
12                                             396940                                                                     
13                                             Jan-19                                                                     
14                                                NaN                                                                     
15                                                NaN                                                                     
16                                      Mode of study                                                                     
17                                                All                                                                     
18                                                All                                                                     
19                                                All                                                                     

   Unnamed: 2              Unnamed: 3                      Unnamed: 4  \
0         NaN                     NaN                             NaN   
1         NaN                     NaN                             NaN   
2         NaN                     NaN                             NaN   
3         NaN                     NaN                             NaN   
4         NaN                     NaN                             NaN   
5         NaN                     NaN                             NaN   
6         NaN                     NaN                             NaN   
7         NaN                     NaN                             NaN   
8         NaN                     NaN                             NaN   
9         NaN                     Na

<h2>Skipping Summary Rows</h2>
<p>Read the datafile and skip the rows that contain summary information.  Use the head method and output the first 20 rows.  The column names should be:
<ul>    
<li>Academic Year</li>
<li>Mode of study</li>
<li>Sex</li>
<li>Country of HE provider</li>
<li>Classification of First Degree</li>
<li>Number</li>
<li>Percentage</li>
</ul>    
</p>

In [51]:
df = pd.read_csv('degreeClassification.csv',skiprows = 17)
df.head(20)

Academic Year Mode of study  Sex Country of HE provider  \
0        2013/14           All  All                    All   
1        2013/14           All  All                    All   
2        2013/14           All  All                    All   
3        2013/14           All  All                    All   
4        2013/14           All  All                    All   
5        2013/14           All  All                England   
6        2013/14           All  All                England   
7        2013/14           All  All                England   
8        2013/14           All  All                England   
9        2013/14           All  All                England   
10       2013/14           All  All       Northern Ireland   
11       2013/14           All  All       Northern Ireland   
12       2013/14           All  All       Northern Ireland   
13       2013/14           All  All       Northern Ireland   
14       2013/14           All  All       Northern Ireland   
15       2013/14           All  All               Scotland   
16       2013/14           All  All               Scotland   
17       2013/14           All  All               Scotland   
18       2013/14           All  All               Scotland   
19       2013/14           All  All               Scotland   

   Classification of First Degree  Number Percentage  
0                           First   79410        20%  
1                    Lower second   95965        24%  
2                      Third/Pass   21905         6%  
3                Total classified  395605        NaN  
4                    Upper second  198320        50%  
5                           First   68590        20%  
6                    Lower second   81665        24%  
7                      Third/Pass   19385         6%  
8                Total classified  339170        NaN  
9                    Upper second  169530        50%  
10                          First    1590        19%  
11                   Lower second    2215        26%  
12                     Third/Pass     285         3%  
13               Total classified    8560        NaN  
14                   Upper second    4470        52%  
15                          First    5125        21%  
16                   Lower second    5505        22%  
17                     Third/Pass     875         4%  
18               Total classified   24880        NaN  
19                   Upper second   13380        54%

<h2>Unique values in a Column</h2>
<p>Output the unique values for the column 'Mode of study' to give a general idea of the categorical data for this column</p>

In [52]:
df[['Mode of study']]

Mode of study
0              All
1              All
2              All
3              All
4              All
...            ...
1375     Part-time
1376     Part-time
1377     Part-time
1378     Part-time
1379     Part-time

[1380 rows x 1 columns]

### as we can see the types of Mode of study are All and Part-time

<h2>Boolean Masks</h2>

<p>Create a boolean mask and then filter the data based on the categorical data value 'All' for 'Mode of Study'.  We do this in stages to help understand how the more complex aggregated instruction is built.</p>
<p>The overall goal is to apply a mask so that all the rows that contain values part-time or full-time are dropped from the dataframe.  We do this so that we don’t double count, i.e. A count in 'All' will also also be a count in either 'Part-time' or 'Full-time'.</p>

<ol>
<li>Create a mask to output True/False for each row, True if 'Mode of study' is equal to 'All' and False otherwise</li>

In [56]:
pip install numpy 

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'C:\Users\asus\Desktop\Cs works\13212-Data-Science\venv\Scripts\python.exe -m pip install --upgrade pip' command.


In [57]:
import numpy as np

In [58]:
df['Mask'] = np.where(df['Mode of study'] == 'All', 'True', 'False')

In [59]:
df

Academic Year Mode of study    Sex Country of HE provider  \
0          2013/14           All    All                    All   
1          2013/14           All    All                    All   
2          2013/14           All    All                    All   
3          2013/14           All    All                    All   
4          2013/14           All    All                    All   
...            ...           ...    ...                    ...   
1375       2017/18     Part-time  Other                England   
1376       2017/18     Part-time  Other                England   
1377       2017/18     Part-time  Other                England   
1378       2017/18     Part-time  Other                England   
1379       2017/18     Part-time  Other                England   

     Classification of First Degree  Number Percentage   Mask  
0                             First   79410        20%   True  
1                      Lower second   95965        24%   True  
2                        Third/Pass   21905         6%   True  
3                  Total classified  395605        NaN   True  
4                      Upper second  198320        50%   True  
...                             ...     ...        ...    ...  
1375                          First       0         ..  False  
1376                   Lower second       0         ..  False  
1377                     Third/Pass       0         ..  False  
1378               Total classified       5        NaN  False  
1379                   Upper second       0         ..  False  

[1380 rows x 8 columns]

### 课堂小笔记：

when creating a new column in dataframe with condition, use np.where("condition", case1, case2)

<ol start='2'>
<li>Assign the Boolean mask to a variable named mask</li>

In [67]:
df

Academic Year Mode of study    Sex Country of HE provider  \
0          2013/14           All    All                    All   
1          2013/14           All    All                    All   
2          2013/14           All    All                    All   
3          2013/14           All    All                    All   
4          2013/14           All    All                    All   
...            ...           ...    ...                    ...   
1375       2017/18     Part-time  Other                England   
1376       2017/18     Part-time  Other                England   
1377       2017/18     Part-time  Other                England   
1378       2017/18     Part-time  Other                England   
1379       2017/18     Part-time  Other                England   

     Classification of First Degree  Number Percentage   Mask  
0                             First   79410        20%   True  
1                      Lower second   95965        24%   True  
2                        Third/Pass   21905         6%   True  
3                  Total classified  395605        NaN   True  
4                      Upper second  198320        50%   True  
...                             ...     ...        ...    ...  
1375                          First       0         ..  False  
1376                   Lower second       0         ..  False  
1377                     Third/Pass       0         ..  False  
1378               Total classified       5        NaN  False  
1379                   Upper second       0         ..  False  

[1380 rows x 8 columns]

<ol start='3'>
<li>Use the variable named mask to update the dataframe named df so that the dataframe is filtered based on the boolean mask</li>
<li>Output the unique values for the column 'Mode of study' to ensure that 'Full-time' and 'Part-time' have been filtered out (don't exist in the dataframe)</li>

In [74]:
df_new=df.drop(columns=['Mode of study'] )

In [75]:
df_new

Academic Year    Sex Country of HE provider  \
0          2013/14    All                    All   
1          2013/14    All                    All   
2          2013/14    All                    All   
3          2013/14    All                    All   
4          2013/14    All                    All   
...            ...    ...                    ...   
1375       2017/18  Other                England   
1376       2017/18  Other                England   
1377       2017/18  Other                England   
1378       2017/18  Other                England   
1379       2017/18  Other                England   

     Classification of First Degree  Number Percentage   Mask  
0                             First   79410        20%   True  
1                      Lower second   95965        24%   True  
2                        Third/Pass   21905         6%   True  
3                  Total classified  395605        NaN   True  
4                      Upper second  198320        50%   True  
...                             ...     ...        ...    ...  
1375                          First       0         ..  False  
1376                   Lower second       0         ..  False  
1377                     Third/Pass       0         ..  False  
1378               Total classified       5        NaN  False  
1379                   Upper second       0         ..  False  

[1380 rows x 7 columns]

In [76]:
df_new.rename(columns={'Mask':'Mode of study'})

Academic Year    Sex Country of HE provider  \
0          2013/14    All                    All   
1          2013/14    All                    All   
2          2013/14    All                    All   
3          2013/14    All                    All   
4          2013/14    All                    All   
...            ...    ...                    ...   
1375       2017/18  Other                England   
1376       2017/18  Other                England   
1377       2017/18  Other                England   
1378       2017/18  Other                England   
1379       2017/18  Other                England   

     Classification of First Degree  Number Percentage Mode of study  
0                             First   79410        20%          True  
1                      Lower second   95965        24%          True  
2                        Third/Pass   21905         6%          True  
3                  Total classified  395605        NaN          True  
4                      Upper second  198320        50%          True  
...                             ...     ...        ...           ...  
1375                          First       0         ..         False  
1376                   Lower second       0         ..         False  
1377                     Third/Pass       0         ..         False  
1378               Total classified       5        NaN         False  
1379                   Upper second       0         ..         False  

[1380 rows x 7 columns]

<ol start='5'>
<li>Output the dataframe</li>

In [77]:
df_new

Academic Year    Sex Country of HE provider  \
0          2013/14    All                    All   
1          2013/14    All                    All   
2          2013/14    All                    All   
3          2013/14    All                    All   
4          2013/14    All                    All   
...            ...    ...                    ...   
1375       2017/18  Other                England   
1376       2017/18  Other                England   
1377       2017/18  Other                England   
1378       2017/18  Other                England   
1379       2017/18  Other                England   

     Classification of First Degree  Number Percentage   Mask  
0                             First   79410        20%   True  
1                      Lower second   95965        24%   True  
2                        Third/Pass   21905         6%   True  
3                  Total classified  395605        NaN   True  
4                      Upper second  198320        50%   True  
...                             ...     ...        ...    ...  
1375                          First       0         ..  False  
1376                   Lower second       0         ..  False  
1377                     Third/Pass       0         ..  False  
1378               Total classified       5        NaN  False  
1379                   Upper second       0         ..  False  

[1380 rows x 7 columns]

<h2>Boolean Masks Continued</h2>

<p>Create a boolean mask and then filter the data based on the categorical data 'All' for 'Sex'</p>
<p>Then create a boolean mask and then filter the data based on the categorical data 'All' for 'Country of HE provider'</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 1380 rows to 25 rows.</p>

In [ ]:
df_new['mask_sex']=np.where(df_new['Sex']=='All',True,False)
df_new['mask_CHP']=np.where(df_new['Country of HE provider']=='All',True,False)

In [ ]:
df_new

In [ ]:
df1=df_new.query('mask_sex'=='True' and 'mask_CHP'=='True')

<h2>More Boolean Masks</h2>

<p>We are not interested in rows that show the totals of classification for each year, filter these out.</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 25 rows to 20 rows.</p>

In [84]:
df1=df_new.query('mask_sex'=='True' and 'mask_CHP'=='True')

ValueError: expr must be a string to be evaluated, <class 'bool'> given

<h2>More Boolean Masks</h2>

<p>We are not interested in rows that show the totals of classification for each year, filter these out.</p>
<p>Output the dataframe to check the filtering was successful, you should have reduced the data from 25 rows to 20 rows.</p>

<h2>Dropping Columns</h2>
<p>Since columns 'Mode of study', 'Sex', and 'Country of HE provider' don\t provide us with any meaninful information (the all contain the data 'All'), drop these columns from the dataframe.</p>
<p>Output the dataframe to check the columns have been dropped successfully.</p>

<h2>Checking the Data Type of a Column</h2>
<p>Check the type of data that is stored for one of the values in the percentage column.</p>

<h2>Cleaning and converting to an string data type to a number data type</h2>
<p>The percentage column is a string data type because of the % character.  Remove the % character and convert the values for this column to a numeric data type</p>

<h2>New Dataframes</h2>
<p>Often it is useful to create seperate smaller dataframes, create four new dataframes named df_first, df_upp_sec, df_low_sec and df_third, to seperate the different degree classifications.</p>

Example:
df_first would output:
<pre>
      Academic Year   Classification of First Degree	Number	 Percentage
0	    2013/14	       First	                    79410	  20%
275	  2014/15	       First	                    81640	  22%
550	  2015/16	       First	                    88890	  24%
825	  2016/17	       First	                    100945	 26%
1105	 2017/18	       First	                    110475	 28%
</pre>

<h2>Save the Dataframes to CSV files</h2>
<p>Save the dataframes df, df_first, df_upp_sec, df_low_sec and df_third to csv files with filesnames that match the dataframe name.